[DISCUSS] [JDK17] Upgrade JUnit / JavaDoc from 4 to 5.

slfan1989 opened the pull request:
<!--
  Thanks for sending a pull request!
    1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/HADOOP/How+To+Contribute
    2. Make sure your PR title starts with JIRA issue id, e.g., 'HADOOP-17799. Your PR title ...'.
-->

### Description of PR

Upgrading JUnit from 4 to 5 is a key task in the process of upgrading our Hadoop project to JDK 17. This task is somewhat complex due to the large number of unit tests and the high coupling between modules. The good news is that our team members have already submitted several upgrade-related PRs, which will significantly ease the workload for the subsequent upgrades. However, the previous upgrade efforts lacked a unified summary and tracking, making it difficult to monitor progress and manage the modules that still need to be upgraded. Therefore, the purpose of this PR is to centralize the tracking of the entire process of upgrading JUnit from 4 to 5, summarize the progress, address the issues encountered during the upgrade, and discuss key challenges.

I have thoroughly reviewed all of our modules and summarized the current upgrade progress.

- The following modules do not require an upgrade:

```
hadoop-assemblies
hadoop-build-tools
hadoop-client
hadoop-client-api
hadoop-client-check-invariants
hadoop-client-check-test-invariants
hadoop-client-minicluster
hadoop-client-runtime
hadoop-cloud-storage
hadoop-annotations
hadoop-auth-example
hadoop-dist
hadoop-hdfs-native-client
hadoop-hdfs-nfs
hadoop-maven-plugins
hadoop-minicluster
hadoop-project
hadoop-project-dist
hadoop-benchmark
hadoop-dynamometer-dist
hadoop-openstack
hadoop-pipes
hadoop-tools-dist
hadoop-yarn-applications-catalog-docker
hadoop-yarn-registry
hadoop-yarn-site
hadoop-yarn-ui
```

- The following modules need to be upgraded (if a module is completed, I will add the JIRA and PR links):

- [x] hadoop-client-integration-tests
HADOOP-19410. Upgrade JUnit from 4 to 5 in hadoop-client-integration-tests. (#7650)

- [x] hadoop-cos
HADOOP-19411. Upgrade JUnit from 4 to 5 in hadoop-cos.(#7640)

- [x] hadoop-huaweicloud
HADOOP-19412. Upgrade JUnit from 4 to 5 in hadoop-huaweicloud.(#7639)

- [x] hadoop-auth
HADOOP-19414. Upgrade JUnit from 4 to 5 in hadoop-auth.(#7638)

- [x] hadoop-common
HADOOP-19415. Upgrade JUnit from 4 to 5 in hadoop-common.(#7339, #7347, #7355, #7356, #7357, #7419, ##7664, #7667, #7668, #7670, #7671, #7672)

- [x] hadoop-kms
HADOOP-19416. Upgrade JUnit from 4 to 5 in hadoop-kms.(#7637)

- [x] hadoop-minikdc
HADOOP-19417. Upgrade JUnit from 4 to 5 in hadoop-minikdc.(#7636, #7661)

- [x] hadoop-nfs
HADOOP-19418. Upgrade JUnit from 4 to 5 in hadoop-nfs.(#7635)

- [x] hadoop-registry
HADOOP-19419. Upgrade JUnit from 4 to 5 in hadoop-registry. (#7624)

- [x] hadoop-hdfs
HDFS-12431. Upgrade JUnit from 4 to 5 in hadoop-hdfs hdfs.(#7383, #7613, #7626, #7647, #7733, #7739, #7794, #7805, #7818, #7841, #7864, #7873, #7890, #7895, #7908, #7922, #7934, #7941)

- [x] hadoop-hdfs-client
HDFS-17718. Upgrade JUnit from 4 to 5 in hadoop-hdfs-client.(#7563)

- [x] hadoop-hdfs-native-client
HDFS-17827. [JDK17] Upgrade JUnit from 4 to 5 in hadoop-hdfs-native-client. (#7950)

- [x] hadoop-hdfs-httpfs
HDFS-17719. Upgrade JUnit from 4 to 5 in hadoop-hdfs-httpfs.(#7371)

- [x] hadoop-hdfs-rbf
HDFS-17720. Upgrade JUnit from 4 to 5 in hadoop-hdfs-rbf.(#7528, #7531,#7949)

- [x] hadoop-mapreduce-client-app
MAPREDUCE-7418. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-app.(#7350)

- [x] hadoop-mapreduce-client-common
MAPREDUCE-7419. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-common.(#5028)

- [x] hadoop-mapreduce-client-core
MAPREDUCE-7420. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-core.(#7363)

- [x] hadoop-mapreduce-client-hs
MAPREDUCE-7414. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-hs.(#7414)

- [x] hadoop-mapreduce-client-hs-plugins
MAPREDUCE-7413. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-hs-plugins.(#5023)

- [x] hadoop-mapreduce-client-jobclient
MAPREDUCE-7421. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-jobclient.(#7358, #7372)

- [x] hadoop-mapreduce-client-nativetask
MAPREDUCE-7415. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-nativetask.(#7349)

- [x] hadoop-mapreduce-client-shuffle
MAPREDUCE-7416. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-shuffle.(#7416)

- [x] hadoop-mapreduce-client-uploader
MAPREDUCE-7417. Upgrade Junit 4 to 5 in hadoop-mapreduce-client-uploader.(#5019)

- [x] hadoop-mapreduce-examples
MAPREDUCE-7422. Upgrade Junit 4 to 5 in hadoop-mapreduce-examples.(#5029)

- [x] hadoop-aliyun
HADOOP-19421. Upgrade JUnit from 4 to 5 in hadoop-aliyun.(#7634)

- [x] hadoop-archive-logs
HADOOP-19422. Upgrade JUnit from 4 to 5 in hadoop-archive-logs.(#7621)

- [x] hadoop-archives
HADOOP-19423. Upgrade JUnit from 4 to 5 in hadoop-archives.(#7620)

- [x] hadoop-aws
HADOOP-19424. Upgrade JUnit from 4 to 5 in hadoop-aws.(#7752,#7814,#7858,#7951)

- [x] hadoop-azure
HADOOP-19425. Upgrade JUnit from 4 to 5 in hadoop-azure.(#7369, #7653, #7674, #7954)

- [x] hadoop-azure-datalake
HADOOP-19426. Upgrade JUnit from 4 to 5 in hadoop-azure-datalake.(#7652, #7757)

- [x] hadoop-compat-bench
HADOOP-19427. Upgrade JUnit from 4 to 5 in hadoop-compat-bench.(#7619)

- [x] hadoop-datajoin
HADOOP-19430. Upgrade JUnit from 4 to 5 in hadoop-datajoin.(#7618)

- [x] hadoop-distcp
HADOOP-19431. Upgrade JUnit from 4 to 5 in hadoop-distcp.(#7368)

- [x] hadoop-dynamometer-blockgen
HADOOP-19428. Upgrade JUnit from 4 to 5 in hadoop-dynamometer-blockgen.(#7590)

- [x] hadoop-dynamometer-infra
HADOOP-19429. Upgrade JUnit from 4 to 5 in hadoop-dynamometer-infra.(#7588)

- [x] hadoop-dynamometer-workload
HADOOP-19432. Upgrade JUnit from 4 to 5 in hadoop-dynamometer-workload.(#7587)

- [x] hadoop-extras
HADOOP-19433. Upgrade JUnit from 4 to 5 in hadoop-extras.(#7586)

- [x] hadoop-federation-balance
HADOOP-19434. Upgrade JUnit from 4 to 5 in hadoop-federation-balance.(#7580)

- [x] hadoop-fs2img
HADOOP-19435. Upgrade JUnit from 4 to 5 in hadoop-fs2img.(#7579)

- [x] hadoop-gridmix
HADOOP-19436. Upgrade JUnit from 4 to 5 in hadoop-gridmix.(#7578)

- [x] hadoop-kafka
HADOOP-19437. Upgrade JUnit from 4 to 5 in hadoop-kafka.(#7547)

- [x] hadoop-resourceestimator
HADOOP-19438. Upgrade JUnit from 4 to 5 in hadoop-resourceestimator.(#7548)

- [x] hadoop-rumen
HADOOP-19439. Upgrade JUnit from 4 to 5 in hadoop-rumen.(#7552)

- [x] hadoop-sls
HADOOP-19440. Upgrade JUnit from 4 to 5 in hadoop-sls.(#7553)

- [x] hadoop-streaming
HADOOP-19441. Upgrade JUnit from 4 to 5 in hadoop-streaming.(#7554)

- [x] hadoop-yarn-api 
YARN-8943. Upgrade JUnit from 4 to 5 in hadoop-yarn-api

- [x] hadoop-yarn-applications-catalog-webapp
YARN-11336. Upgrade Junit 4 to 5 in hadoop-yarn-applications-catalog-webapp (#4992)

- [x] hadoop-yarn-applications-distributedshell
YARN-11760. Upgrade JUnit from 4 to 5 in hadoop-yarn-applications-distributedshell.(#7373)

- [x] hadoop-yarn-applications-mawo-core
YARN-11337. Upgrade Junit 4 to 5 in hadoop-yarn-applications-mawo(#4993)

- [x] hadoop-yarn-applications-unmanaged-am-launcher
YARN-11338. Upgrade Junit 4 to 5 in hadoop-yarn-applications-unmanaged-am-launcher(#4994)

- [x] hadoop-yarn-services-api
YARN-11339. Upgrade Junit 4 to 5 in hadoop-yarn-services-api.(#4995)

- [x] hadoop-yarn-services-core
YARN-11761. Upgrade JUnit from 4 to 5 in hadoop-yarn-services-core.(#7374)

- [x] hadoop-yarn-client
YARN-11244. Upgrade JUnit from 4 to 5 in hadoop-yarn-client.(#7438)

- [x] hadoop-yarn-common
YARN-6946. Upgrade JUnit from 4 to 5 in hadoop-yarn-common.(#4717)

- [x] hadoop-yarn-csi
YARN-11245. Upgrade JUnit from 4 to 5 in hadoop-yarn-csi (#4778)

- [x] hadoop-yarn-server-applicationhistoryservice
YARN-10793. Upgrade Junit from 4 to 5 in hadoop-yarn-server-applicationhistoryservice(#4603)

- [x] hadoop-yarn-server-common
YARN-11258. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-common.(#7437)

- [x] hadoop-yarn-server-globalpolicygenerator
YARN-11762. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-globalpolicygenerator.(#7389)

- [x] hadoop-yarn-server-nodemanager
YARN-11263. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-nodemanager.(#7390, #7455)

- [x] hadoop-yarn-server-resourcemanager
YARN-11262. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-resourcemanager.(#7474, #7484, #7485, #7488, #7489, #7490)

- [x] hadoop-yarn-server-router
YARN-11267. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-router.(#7546)

- [x] hadoop-yarn-server-sharedcachemanager
YARN-11265. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-sharedcachemanager.(#4772)

- [x] hadoop-yarn-server-tests
YARN-11264. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-tests.(#4776)

- [x] hadoop-yarn-server-timeline-pluginstorage
YARN-11269. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-timeline-pluginstorage.(#4771)

- [x] hadoop-yarn-server-timelineservice
YARN-11260. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-timelineservice.(#4775)

- [x] hadoop-yarn-server-timelineservice-documentstore
YARN-11268. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-timelineservice-documentstore.(#7450)

- [x] hadoop-yarn-server-timelineservice-hbase-client
YARN-11270. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-timelineservice-hbase-client.(#4773)

- [x] hadoop-yarn-server-timelineservice-hbase-common
YARN-11271. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-timelineservice-hbase-common. (#4774)

- [x] hadoop-yarn-server-timelineservice-hbase-server-2
YARN-11763. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-timelineservice-hbase-server-2.

- [x] hadoop-yarn-server-timelineservice-hbase-tests
YARN-11266. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-timelineservice-hbase-tests. (#7538)

- [x] hadoop-yarn-server-web-proxy
YARN-11261. Upgrade JUnit from 4 to 5 in hadoop-yarn-server-web-proxy.(#4777)


- Tasks to complete before releasing Hadoop 3.5.0:

- [] YARN has 4 unit tests marked as "Disabled" that need further follow-up. (#7345)

- [] The YARN RESTful APIs require additional validation to ensure compatibility. (#7453)

### How was this patch tested?


### For code changes:

- [ ] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

slfan1989 committed:
[JDK17] Upgrade JUnit from 4 to 5 in hadoop-common.

hadoop-yetus commented:
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m  0s |  |  Docker mode activated.  |
| -1 :x: |  patch  |   0m 21s |  |  https://github.com/apache/hadoop/pull/7337 does not apply to trunk. Rebase required? Wrong Branch? See https://cwiki.apache.org/confluence/display/HADOOP/How+To+Contribute for help.  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-7337/1/console |
| versions | git=2.34.1 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

jojochuang commented:
Wow! Appreciate such a huge work!
You'd probably find it easier to divide it into subtasks, component by component. This is impossible to review too.

slfan1989 commented:
> Wow! Appreciate such a huge work!
> You'd probably find it easier to divide it into subtasks, component by component. This is impossible to review too.

Thank you for your message! This PR is currently an experimental version (it cannot be merged directly). I have attempted to upgrade some modules related to hadoop-common from JUnit 4 to JUnit 5. During the upgrade process, there are some issues that need to be discussed with everyone to establish the basic standards for this module upgrade.

slfan1989 commented:
@steveloughran @ayushtkn @Hexiaoqiao @jojochuang @aajisaka @cnauroth @szetszwo 

Upgrading JUnit from 4 to 5 is not only a technical upgrade opportunity but also a good chance for us to improve the test-related code and fix potential issues. I’d like to discuss whether the improvement strategies I’ve been considering are reasonable.

1. **Import Optimization**

There are currently two ways to reference JUnit methods in the project:

- The first way is by importing static methods, allowing us to directly use assertTrue in unit tests.
- The second way is by importing `org.junit.Assert`, which requires us to use the fully qualified name (e.g., Assert.assertTrue) in unit tests.

I hope to unify all references to the first method (static imports) in this update, but I'm unsure if this is feasible.

2. **Import * Optimization**

I’ve noticed that many unit tests use the `import *` approach. Do we need to expand these imports in the unit tests?

3. **Addressing Checkstyle Issues, Incomplete JavaDocs, and Typos**

There are several Checkstyle issues in the code, such as incomplete JavaDocs and typos. Should we also address these issues in this update? 

Some logs use `+` for string concatenation instead of placeholders. Some methods could also be optimized using lambda expressions. Should we address these issues in this update as well?

4. **Improvements for Large Modules**

For example, the `hadoop-common` module, I’ve noticed, has many downstream dependencies, affecting several other modules, as shown below:

```
hadoop-azure
hadoop-aws
hadoop-distcp
hadoop-azure-datalake
hadoop-hdfs
hadoop-hdfs-rbf
hadoop-cos
hadoop-aliyun
hadoop-mapreduce-client-core
```
We can’t make all the changes at once because there are too many involved modules, and some test classes have dependencies on each other.

I’ve found that JUnit 4 and JUnit 5 can coexist, meaning that a module can have both JUnit 4 and JUnit 5 test styles, and the tests will still run successfully. For very large modules, we can modify them incrementally, based on the package structure. For example, for `hadoop-common`, we can submit multiple PRs to complete the module’s upgrade step by step. Make sure that each PR submitted doesn’t include changes to more than 50 classes, as this will make the review process easier.

HADOOP-19415. [JDK17] Upgrade JUnit from 4 to 5 in hadoop-common Part1.
```
cli、conf、constants、crypto、fs
```

HADOOP-19415. [JDK17] Upgrade JUnit from 4 to 5 in hadoop-common Part2.
```
ha、http
```

If we encounter unit tests that depend on other modules, we can leave them unchanged for now. Once the non-dependent parts of the other modules are upgraded, we can submit a single PR to complete the upgrade of the dependencies.

Let’s illustrate with an example:

The `FCStatisticsBaseTest` class in `hadoop-common` has dependencies on the following three modules:  
`hadoop-aws`, `hadoop-common`, and `hadoop-aliyun`.

<img width="825" alt="image" src="https://github.com/user-attachments/assets/c6ea830a-3f57-4b95-8a01-3fffbcb8712b" />

Due to these dependencies, we will not upgrade this abstract class for now. Instead, we will parallelly upgrade the parts of these three modules(`hadoop-aws`, `hadoop-common`, and `hadoop-aliyun`) that have no dependencies. Once the other parts are upgraded, we will come back to upgrade the unit tests with dependencies.

I’d also like to sincerely thank @zhtttylz for providing the unit test upgrade tool. With this tool, I was able to complete the PR, which involves multiple modules, in just two days, significantly improving efficiency.  If we can confirm the coding guidelines, @zhtttylz and I will submit the relevant PRs together to complete the JUnit version upgrade. 

I’m also willing to volunteer to drive and follow up on the upgrade work to ensure it’s completed in a shorter time frame. I estimate the upgrade should take about 1-3 months.

cnauroth commented:
Wow, this is awesome @slfan1989 and @zhtttylz ! Here are my opinions on how to proceed.

> 1. **Import Optimization**
> ...
> 2. **Import * Optimization**
> ...
> 3. **Addressing Checkstyle Issues, Incomplete JavaDocs, and Typos**
> ...

I'd like to suggest that in this phase, we stick to a minimal effort of making a straight translation to JUnit 5 in the interest of completing a JDK 17 release as soon as possible. These are going to be large pull requests to review, even after we break them into sub-tasks. It will take longer to code review if we start including other unrelated improvements.

All of these proposals are good code quality improvements though. We can pick them up as future work after completing the JUnit 5 migration and after the JDK 17 release.

I've also heard there's a desire to go toward AssertJ. If so, then that might change the plan on points 1 and 2.

> 4. **Improvements for Large Modules**
> ...
> Due to these dependencies, we will not upgrade this abstract class for now. Instead, we will parallelly upgrade the parts of these three modules(`hadoop-aws`, `hadoop-common`, and `hadoop-aliyun`) that have no dependencies. Once the other parts are upgraded, we will come back to upgrade the unit tests with dependencies.

I agree with your strategy here. It sounds like this is the best way to manage pull request size and code review process.

slfan1989 commented:
> Wow, this is awesome @slfan1989 and @zhtttylz ! Here are my opinions on how to proceed.
> 
> > 1. **Import Optimization**
> >    ...
> > 2. **Import * Optimization**
> >    ...
> > 3. **Addressing Checkstyle Issues, Incomplete JavaDocs, and Typos**
> >    ...
> 
> I'd like to suggest that in this phase, we stick to a minimal effort of making a straight translation to JUnit 5 in the interest of completing a JDK 17 release as soon as possible. These are going to be large pull requests to review, even after we break them into sub-tasks. It will take longer to code review if we start including other unrelated improvements.
> 
> All of these proposals are good code quality improvements though. We can pick them up as future work after completing the JUnit 5 migration and after the JDK 17 release.
> 
> I've also heard there's a desire to go toward AssertJ. If so, then that might change the plan on points 1 and 2.
> 
> > 4. **Improvements for Large Modules**
> >    ...
> >    Due to these dependencies, we will not upgrade this abstract class for now. Instead, we will parallelly upgrade the parts of these three modules(`hadoop-aws`, `hadoop-common`, and `hadoop-aliyun`) that have no dependencies. Once the other parts are upgraded, we will come back to upgrade the unit tests with dependencies.
> 
> I agree with your strategy here. It sounds like this is the best way to manage pull request size and code review process.

@cnauroth Thank you for your message! I agree with your point of view. We will only make the necessary changes and will not adjust other improvements for now. Once the JUnit 5 upgrade is complete, we will proceed with further optimizations.

Here are the improvement principles I've summarized:

1. Focus on making the necessary changes to ensure the upgrade is completed smoothly.
2. If the Yetus report highlights Checkstyle issues or other problems, we will address them to achieve a successful compile result.
3. Break down large modules into smaller PRs to make the code review process more efficient and user-friendly.

cc: @steveloughran @ayushtkn @Hexiaoqiao

cnauroth commented:
@slfan1989 , great summary. Thank you for these patches!

slfan1989 commented:
After the merge of HADOOP-15984, we noticed that some unit tests stopped running.  To resolve this, we found a solution that allows the unit tests to run again.  We can refer to HADOOP-19405. hadoop-aws and hadoop-azure tests have stopped running.(#7335) for more details. 

However, during the process of completing this PR, a new issue was exposed — we encountered a large number of Javadoc errors. During the JDK 11 upgrade, we addressed issues in some modules, but many modules still have Javadoc problems. We plan to address these issues during the JDK 17 upgrade process.

I do not intend to create a separate JIRA and PR for each module to fix the Javadoc errors, as this would generate significant noise. Improving Javadoc does not pose any execution risks since the PRs won’t modify the code itself.

For larger modules, such as hadoop-common, hdfs, hdfs-rbf, resourcemanager, and nodemanager, we will create individual JIRAs. For other modules, we will submit the changes under larger module scopes, such as `hadoop-tools`, `yarn`, `mapreduce`, and `hdfs`.

slfan1989 commented:
To sync the recent progress, I have completed the upgrade of all MapReduce module tests from JUnit4 to JUnit5 locally (there may still be some issues, but we'll wait for Yetus feedback). I plan to start submitting the PRs tomorrow.

slfan1989 commented:
@cnauroth @steveloughran 

The following PR is ready, please help to review it

MapReduce:
MAPREDUCE-7420. [JDK17] Upgrade Junit 4 to 5 in hadoop-mapreduce-client-core Part1. #7363
MAPREDUCE-7421. [JDK17] Upgrade Junit 4 to 5 in hadoop-mapreduce-client-jobclient Part1. #7358

HADOOP:
HADOOP-19431. [JDK17] Upgrade JUnit from 4 to 5 in hadoop-distcp. #7368
HADOOP-19425. [JDK17] Upgrade JUnit from 4 to 5 in hadoop-azure Part1. #7369

HDFS:
HDFS-17719. [JDK17] Upgrade JUnit from 4 to 5 in hadoop-hdfs-httpfs Part1. #7371

slfan1989 commented:
I’d like to provide an update on the progress of the JUnit 5 upgrade and will share periodic updates and specific details:

The MapReduce module upgrade has been completed.

For the YARN module upgrade, the remaining modules include:

- hadoop-yarn-server-resourcemanager
- hadoop-yarn-server-router
- hadoop-yarn-server-timelineservice-hbase-tests

The related improvement PRs have been submitted (compilation is mostly complete), but there are still some issues (e.g., checkstyle problems and failing unit tests), which are being actively addressed.

The hadoop-common module upgrade has progressed to Part 6. After Part 6, there will be several more PRs, most of which involve cross-module changes requiring collaboration from different module members for code reviews. I plan to continue pushing these PRs forward next week.

My expectation is to complete the entire JUnit 5 upgrade by the end of March to mid-April. Since time is tight, I’ve invited @zhtttylz  to assist with the upgrade. He will be responsible for helping upgrade the HDFS-related unit tests, and this work will begin next week.

cc: @ayushtkn @cnauroth @steveloughran @Hexiaoqiao

h-vetinari commented:
Is there a reason why all these tasks (resp. this PR) aren't linked to https://issues.apache.org/jira/browse/HADOOP-14693? It makes it a bit hard to follow the current status/progress of the conversion.

slfan1989 commented:
We have made some progress in upgrading from Junit4 to Junit5. I will summarize the current progress and share the relevant updates by the end of this week.

slfan1989 commented:
> s there a reason why all these tasks (resp. this PR) aren't linked to https://issues.apache.org/jira/browse/HADOOP-14693? It makes it a bit hard to follow the current status/progress of the conversion.

@h-vetinari Thank you for your question! I will provide a complete update on the progress of the Junit4 to Junit5 upgrade this week and link the relevant JIRA tasks.

slfan1989 commented:
I’d like to provide an update on the progress of upgrading Hadoop from JUnit 4 to JUnit 5. I’ve outlined all the modules that need to be upgraded, totaling 51 modules. Currently, we have completed the upgrade of 42 modules, with 9 modules still in progress, bringing the overall progress to over 80%.

Here’s the detailed progress:

HADOOP-19414: Upgrade JUnit from 4 to 5 in hadoop-auth (#7638)
The PR has been created, and local unit tests have passed. We are now waiting for feedback from Yetus, and some checkstyle issues may still need to be addressed.

HADOOP-19415: Upgrade JUnit from 4 to 5 in hadoop-common
This module still requires 3-5 PRs to fix unit tests with inheritance relationships. #7419 is still being followed up, and once ready, I’ll invite community members for review.

HDFS-12431: Upgrade JUnit from 4 to 5 in hadoop-hdfs
This module may require at least 10 more PRs, and @zhtttylz  is currently following up on it.

HADOOP-19421: Upgrade JUnit from 4 to 5 in hadoop-aliyun (#7634)
The PR is ready and waiting for review. We need a team member familiar with this module to help with the review.

HADOOP-19433: Upgrade JUnit from 4 to 5 in hadoop-extras (#7586)
The PR is ready and waiting for review.

hadoop-aws, hadoop-azure, hadoop-azure-datalake
These modules are progressing in parallel, with no PRs submitted yet. The upgrade is expected to start next week.

slfan1989 commented:
I am continuing to work on the JUnit5 upgrade for the Hadoop-common module. The remaining tasks are more challenging, involving cross-module unit test upgrades. The good news is that I have mostly completed the task of splitting the remaining unit tests, specifically for PART6 to PART12, and have finished one round of dependency upgrades. I estimate that it will take about another week to complete the development work for PART6 to PART12.

slfan1989 commented:
@cnauroth @steveloughran @Hexiaoqiao @ayushtkn @szetszwo @anujmodi2021 @jojochuang @mukund-thakur

The Hadoop Common module has been split into Part6 to Part12 and is now ready for review. I greatly need your help to continue advancing the Junit5 upgrade work.

Part6 PR (#7419) primarily modifies the unit tests of AbstractFSContractTestBase and its subclasses, involving the hadoop-common, hadoop-hdfs, hadoop-mapreduce, hadoop-aws, hadoop-azure, and hadoop-azure-datalake modules. This part cannot be split, and I kindly ask for your review.

Part7 PR (#7664) primarily modifies certain classes related to dependencies in the hadoop-common and hadoop-hdfs modules.

Part8 PR (#7667) primarily modifies the unit tests of SymlinkBaseTest, FCStatisticsBaseTest, FileContextCreateMkdirBaseTest, and their subclasses, involving the hadoop-common and hadoop-hdfs modules.

Part9 PR (#7668) primarily modifies the unit tests of HadoopTestBase and its subclasses, involving the hadoop-common, hadoop-hdfs, hadoop-mapreduce, hadoop-aws, hadoop-azure, and hadoop-azure-datalake modules.

Part10 PR (#7670) primarily modifies the unit tests of AbstractAuditingTest, AbstractHadoopTestBase, and their subclasses, involving the hadoop-common, hadoop-hdfs, hadoop-mapreduce, hadoop-aws, hadoop-azure, and hadoop-azure-datalake modules.

Part11 PR (#7671) primarily modifies the IO-related unit tests in the hadoop-common module.

Part12 PR (#7672) primarily optimizes the IO-related unit tests in the hadoop-common module.

I hope everyone can provide valuable feedback to help us smoothly move forward with the next steps. Thank you!

slfan1989 commented:
Things have been a bit busy over the past two weeks, so the progress on JUnit 5 was delayed. I’ll be picking it up again this week and will continue to push it forward.

slfan1989 commented:
Here's an update on the progress of the JUnit 5 migration. Over the past month, due to other priorities, the pace has been a bit slower, but we've still made meaningful progress. Most modules have been successfully upgraded, with the exception of the following:

- Hadoop-Common: Three PRs are still pending. Among them, #7664 and #7670 are ready to be review, while #7419 requires further refinement.
- HDFS: At least 10 PRs are yet to be submitted. @zhtttylz  is currently preparing them. Once the Hadoop-Common changes are merged, the pace of HDFS submissions will accelerate.
- Hadoop-Azure: Two PRs are in progress and will be resumed after Hadoop-Common is merged.
- Hadoop-AWS: This module also depends on the Hadoop-Common merge and will proceed accordingly.
- YARN-Router: Two PRs are needed to complete the upgrade for this module.

In the next two weeks, I’ll have more availability and plan to complete the remaining dependency upgrades for all modules except HDFS.

cc: @cnauroth @Hexiaoqiao @zhtttylz

slfan1989 commented:
The JUnit5 upgrade for the Hadoop-common module has been completed, and the upgrade code for the AWS and Yarn-router modules is ready. The Azure module is still being worked on, with two PRs expected. I am currently attempting to remove the JUnit4 dependency locally and have encountered a few minor issues, which are being addressed. The goal is to complete the JUnit5 upgrade for all modules, except HDFS, by the end of June 2025.

cc: @cnauroth @steveloughran @Hexiaoqiao @zhtttylz @h-vetinari

slfan1989 commented:
Here’s an update on the progress of the JUnit 5 upgrade:

The hadoop-aws module has completed HADOOP-19424 and removed the JUnit 4 dependency. Steve will provide further guidance in the future, and we plan to continue upgrading hadoop-aws to use a higher version of JUnit.

The hadoop-azure-datalake module is nearly finished with the upgrade.

The hadoop-yarn-server-router module has completed the upgrade.

The hadoop-azure and hadoop-hdfs modules are expected to take some more time.

This week, I will submit the final version of the hadoop-azure part3 upgrade code and begin working on removing the JUnit 4 dependency from the mapreduce, yarn, and tools modules.

slfan1989 commented:
Currently, except for the HDFS module, the other modules (including Common, Cloud, Tools, Yarn, and Mapreduce) have been upgraded, and the Junit4 dependencies have been completely removed. Next, I will conduct a thorough check of these modules within the next 1-2 days to confirm if there are any areas that need further improvement, though it is expected that there won't be many. As for the HDFS module, I have had offline discussions with @zhtttylz , and there are still 5-6 PRs being processed. At the same time, I will first try to remove the Junit4 dependencies from other submodules within the HDFS module.

cc: @ayushtkn @cnauroth @Hexiaoqiao @h-vetinari @steveloughran @zhtttylz

slfan1989 commented:
Let me share an update on the latest progress of the JUnit 5 upgrade in Hadoop: this week, the JUnit 5 migration for the HDFS module has been successfully completed — special thanks to @zhtttylz for the strong support!

I'm currently working locally to completely remove the dependency on JUnit 4 from the trunk branch. During the process, I found that a small number of unit tests still haven’t been upgraded, and I’m actively working on addressing them. Hopefully, the entire upgrade can be fully wrapped up by the end of this week.

cc: @ayushtkn @cnauroth @Hexiaoqiao @h-vetinari @steveloughran @zhtttylz

slfan1989 commented:
I would like to share the current progress on Hadoop JDK17.

1. Junit5 Upgrade

   We have completed the migration of unit tests to Junit5 across the project and removed Junit4 dependencies wherever possible.   In addition, we configured an Import Rule to ensure that Junit4 references are not reintroduced. However, some Junit4 dependencies still remain because certain third-party plugins continue to rely on Junit4 components (
see YARN-11872: TestAppCatalogSolrClient has a Junit4 transitive dependency). We will need to wait until these plugins are upgraded before fully removing Junit4.

2. Unit Test Issues
  Several failing unit tests remain, mainly in the YARN module, including RM crash tests and other errors.
  There are also issues reported with the Jesery2 Client (see [mailing list thread](https://lists.apache.org/thread/y70j1wqj00rhon4hxjkws6k55514sbfo)).
  I have set up a complete local environment to run the full test suite and will continue to follow up on these issues with high priority.

3. JDIFF Issues and Progress
  JDIFF is a long-standing plugin that has not been updated since JDK8.  Thanks to the efforts of Hualong Zhang, JDIFF can now be successfully compiled under JDK17.  He also updated the custom annotations in hadoop-annotations, enabling jdiff to generate difference files under JDK17. I assisted in reviewing part of the code changes and consider them reasonable.  Hualong is currently on leave and is expected to submit these changes to the community next week.
Additionally, we identified missing files in some versions:

```
  hadoop-3.4.0 is missing Apache_Hadoop_Common_3.4.0.xml;
  hadoop-3.4.1 and hadoop-3.4.2 did not upload jdiff comparison files to the trunk branch.
```

4. JDK17 Validation Pipeline

   With the assistance of Gautham, we have configured a Jenkins JDK17 validation pipeline.   I will run a full validation on the trunk branch as soon as possible, and if no issues are found,   we plan to officially switch the trunk branch build to JDK17 within the next 1–2 weeks.

slfan1989 commented:
- The ResourceManager crash issue has been fixed (#8014).
- The TestUberAM issue has been resolved.  (#8017)
- The issue with Jersey 2.x client is currently being addressed.

I am running a full set of unit tests locally and hope the results will meet expectations.
